src/wp-includes/webfonts-api/class-wp-webfonts-schema-validator.php

```diff
@@ -0,0 +1,330 @@
+<?php
+/**
+ * Webfonts API: Webfonts Schema Validator
+ *
+ * @package WordPress
+ * @subpackage Webfonts
+ * @since 5.9.0
+ */
+
+/**
+ * Webfonts Schema Validator.
+ *
+ * Validates the webfont schema.
+ *
+ * @since 5.9.0
+ */
+class WP_Webfonts_Schema_Validator {
```

Comment:
>Looking again at WP_Webfonts_Schema_Validator, still don't see the point of having it at this place. Would be great if its existence is confirmed by providing some examples of how it is supposed to work, and when.

@azaozz The Validator exists first and foremost validate required keys exist and their value is of the right data type. Why?  To avoid PHP 8.1 deprecation notices for wrong data types, fatal errors from missing required keys, and future with PHP 9 fatal errors from wrong data types.

Here are the methods that are strictly checking those required elements:

- `is_valid_provider()`
- `is_valid_font_family()`
- `is_src_valid()`

Notice that in each of these, if they are not valid, `trigger_error()` happens with a message to help extenders debug their code and resolve schema issues.

>For the lower level functionality WordPress has a well established way to set defaults when an array is used as function parameter: wp_parse_args() https://developer.wordpress.org/reference/functions/wp_parse_args/. It is used pretty much everywhere in the same case as in this patch. Frankly I don't see why it is not used here.

Yes, `wp_parse_args()` can set defaults if elements (via key) don't already exist. However, if the elements exist in the given webfont, `wp_parse_args()` does not overwrite them.

This can be handy to set required keys. I did consider using it to set the base schema structure. However, as noted above, these required elements must also be the correct data type and format to avoid deprecation notices and fatal errors.

>Frankly I've never seen an internal, low level API to have a "schema" and "validation" of that schema. 

I can appreciate that. These types of validations are throughout core, but embedded within the code that's doing the processing. We do things like check if a key exists, if it's empty or not, if it's the right data type or not, etc.

Here, I chose to break it out on its own because it has a specific purpose. It could have been embedded inside of the `WP_Webfonts_Registry`. But then it makes that class big and less readable. Plus, the 2 classes have different jobs and reasons to exist.

For the performance side, this validator runs in microseconds. I ran benchmarks and published that earlier in the thread. It's super duper fast.

There is a case to be made against checking the actual value of each CSS property. The concern was the impact on users as some invalid values would result in their fonts not being fetched from the font service and/or not rendering correctly. By checking them in this Validator, we are going a step further to ensure users fonts display properly.

What do you think?

Reply:
Replying to https://github.com/WordPress/wordpress-develop/pull/1736#issuecomment-962024950.

If I'm understanding you correctly @azaozz , you're proposing to:

- Add a global function that calls the Validator
- Shift calling the Validator to themes and plugins and don't run it internally within the API
- Make the API fully trust the incoming webfonts collections

Is this correct?

What happens if a plugin or theme does not validate and it causes errors or unexpected styling for users?

Reply:
Hey @felixarntz @aristath What do you think about @azaozz proposal https://github.com/WordPress/wordpress-develop/pull/1736#issuecomment-962024950 for the validation?

Reply:
Replying to @azaozz's https://github.com/WordPress/wordpress-develop/pull/1736#issuecomment-962044838

There are a few ways to test what will happen:
* Directly in a URL to the Google Fonts API
* Commenting out the Validator in the `WP_Webfonts_Registry::register()` so it never runs

I did some testing to find out what would happen. [Here's my test report](https://gist.github.com/hellofromtonya/70dc940b27387c46fbaa3d3794e2bbd8).

tl;dr
* Invalid `font-family` or `font-weight` in a 'google' webfont results in a request error
     * e.g. [invalid font-family URL](https://fonts.googleapis.com/css2?family=Invalid:wght@400) and [invalid font-weight URL](https://fonts.googleapis.com/css2?family=Open+Sans:wght@invalid)
     * Impact on the website: `@font-face` is not generated. Font will not display
* Local: non-string `font-family`, `font-style`, `font-weight`, etc. 
     * Results: the property not being included in the `@font-face` style
     * Depending upon the PHP version: notice, warning, or fatal error
     * Impact on the website: font will not display or displays incorrectly

Guarding against missing keys and non-string values should be included.

But for validating the actual string value, the question becomes: Should the Webfonts API protect against invalid `font-style`, `font-weight`, and `font-display` values? Please note: currently, an invalid value is set to a default value to ensure the `@font-face` generates. Should it?